bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceRegistrationImpl.java

```diff
@@ -535,28 +537,54 @@ S getService(BundleContextImpl user, ServiceConsumer consumer) {
 			}
 
 			/* Obtain and return the service object */
-			synchronized (use) {
-				/* if another thread removed the ServiceUse, then
-				 * go back to the top and start again */
-				synchronized (servicesInUse) {
-					user.checkValid();
-					if (servicesInUse.get(this) != use) {
-						continue;
+			ReentrantLockExt useLock = use.getLock();
+			try {
+				if (useLock.tryLock(use.getLockTimeout(), TimeUnit.SECONDS)) {
+					if (registry.debug.DEBUG_SERVICES) {
+						Debug.println('[' + Thread.currentThread().getName() + "] getServiceLock[" //$NON-NLS-1$
+								+ user.getBundleImpl() + "](" + this + "), id:" + System.identityHashCode(useLock) //$NON-NLS-1$ //$NON-NLS-2$
+								+ ". holdCount:" + useLock.getHoldCount() //$NON-NLS-1$
+								+ ", queued:" + useLock.getQueueLength()); //$NON-NLS-1$
 					}
-				}
-				S serviceObject = consumer.getService(use);
-				/* if the service factory failed to return an object and
-				 * we created the service use, then remove the
-				 * optimistically added ServiceUse. */
-				if ((serviceObject == null) && added) {
+
+					/* if another thread removed the ServiceUse, then
+					 * go back to the top and start again */
 					synchronized (servicesInUse) {
-						synchronized (registrationLock) {
-							servicesInUse.remove(this);
-							contextsUsing.remove(user);
+						user.checkValid();
+						if (servicesInUse.get(this) != use) {
+							if (registry.debug.DEBUG_SERVICES) {
+								Debug.println('[' + Thread.currentThread().getName() + "] getServiceContinue[" //$NON-NLS-1$
+										+ user.getBundleImpl() + "](" + this //$NON-NLS-1$
+										+ ")"); //$NON-NLS-1$
+							}
+							continue;
 						}
 					}
+					S serviceObject = consumer.getService(use);
+					/* if the service factory failed to return an object and
+					 * we created the service use, then remove the 
+					 * optimistically added ServiceUse. */
+					if ((serviceObject == null) && added) {
+						synchronized (servicesInUse) {
+							synchronized (registrationLock) {
+								servicesInUse.remove(this);
+								contextsUsing.remove(user);
+							}
+						}
+					}
+					return serviceObject;
+				}
+				throw new ServiceException(
+						"Failed to acquire lock within try period, Lock, id:" + System.identityHashCode(useLock) + ", " //$NON-NLS-1$ //$NON-NLS-2$
+								+ useLock.toString(),
+						ServiceUse.DEADLOCK);
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt(); // To appease sonar
+				return null;
+			} finally {
+				if (useLock.isHeldByCurrentThread()) {
+					useLock.unlock();
```

Comment:
This finally should be part of a try/finally block in the if tryLock block and always call unlock. It is too risky here.